### PR TITLE
Bugfix/styled components

### PR DESCRIPTION
**Pull request by KarlGe**
Fix to #3595 based on the suggestion by @nguyenquyhy 
- Moved all defaultProps "as" to props destructuring instead of defaultProps
- Added tests for *most* instances
- - Tests that defaultProps doesn't contain "as"
- - Tests that the component still renders with the correct default element
- - I Decided to leave the defaultProps even if it's now empty.

For any component that uses createBootstrapComponent I either used or made a const for the returned object and added the defaultProps to that. For InputGroup for example I added the defaultProps to the DecoratedInputGroup, instead of to the InputGroup directly. Thats necessary to check if the class contains "as" in defaultProps. **Though I'm not sure if this is the best way!**

Any class that had the defaultProps set as a static member of the class, has been moved out of the class definition and defined as ClassName.defaultProps = {}

Didn't manage to get those tests to work with Carousel as setting defaultProps on the DecoratedCarousel lead to
` Uncaught Error: Warning: Failed prop type: You have provided a 'activeIndex' prop to 'Carousel' without an 'onSelect' handler prop. This will render a read-only field. If the field should be mutable use 'defaultActiveIndex'. Otherwise, set 'onSelect'.`

Didn't make the tests for AbstractNav either, but maybe that's fine?

### CODE REVIEW

File: src/AbstractNav.js [hunk 1]
```diff
@@ -27,9 +27,7 @@ class AbstractNav extends React.Component {
     activeKey: PropTypes.any,
   };
 
-  static defaultProps = {
-    as: 'ul',
-  };
+  static defaultProps = {};
```
Comment (taion) [inline src/AbstractNav.js]: remove the empty object stubs?

File: src/Jumbotron.js [hunk 2]
```diff
@@ -4,22 +4,27 @@ import classNames from 'classnames';
...
-  bsPrefix: PropTypes.string,
-};
-
 const defaultProps = {
-  as: 'div',
   fluid: false,
 };
 
 class Jumbotron extends React.Component {
+  static propTypes = {
```
Comment (taion) [inline src/Jumbotron.js]: maybe undo the changes to code style for statics?

File: src/Navbar.js [hunk 3]
```diff
@@ -210,13 +209,13 @@ class Navbar extends React.Component {
...
 
 Navbar.propTypes = propTypes;
-Navbar.defaultProps = defaultProps;
 
 const DecoratedNavbar = createBootstrapComponent(
   uncontrollable(Navbar, { expanded: 'onToggle' }),
   'navbar',
 );
 
+DecoratedNavbar.defaultProps = defaultProps;
```
Comment (taion) [inline src/Navbar.js]: why this change?

File: src/Row.js [hunk 4]
```diff
@@ -40,4 +35,10 @@ class Row extends React.Component {
   }
 }
 
-export default createBootstrapComponent(Row, 'row');
+const BootstrapRow = createBootstrapComponent(Row, 'row');
+
+BootstrapRow.defaultProps = {
```
Comment (taion) [inline src/Row.js]: why this change?

File: src/Spinner.js [hunk 5]
```diff
@@ -83,4 +78,8 @@ class Spinner extends React.Component {
   }
 }
 
-export default createBootstrapComponent(Spinner, 'spinner');
+const BootstrapSpinner = createBootstrapComponent(Spinner, 'spinner');
+
+BootstrapSpinner.defaultProps = {};
```
Comment (taion) [inline src/Spinner.js]: remove, probably

File: test/AccordionSpec.js [hunk 6]
```diff
@@ -89,4 +89,12 @@ describe('<Accordion>', () => {
 
     expect(onClickSpy).to.be.calledOnce;
   });
+
+  it('Should define default "as" in prop destructuring instead of deafultProps', () => {
```
Comment (taion) [inline test/AccordionSpec.js]: i'm not sure these test cases really add much. i think it's fine as long as the old tests pass.

File: src/AbstractNav.js (hunk 1 above)
Comment (KarlGe) [reply src/AbstractNav.js]: Yeah, was unsure about that, I can definitely remove them!

File: src/Navbar.js (hunk 3 above)
Comment (KarlGe) [reply src/Navbar.js]: In order to pass the "Should define default "as" in prop destructuring" tests. If those tests aren't necessary then this doesn't have to be there.

File: src/Row.js (hunk 4 above)
Comment (KarlGe) [reply src/Row.js]: Same as Navbar, to pass the test

File: test/AccordionSpec.js (hunk 6 above)
Comment (KarlGe) [reply test/AccordionSpec.js]: You're right, I was mostly thinking about having them in order to prevent someone from inadvertently putting the "as" into the defaultProps at a later point. This whole thing can be considered a bit weird, so I thought this would be an ok way to show why they're set up like they are. What do you think?

Comment (taion) [reply test/AccordionSpec.js]: I think it'd be more effective to add a comment in the places where you set the default, honestly. Otherwise nothing stops someone from carelessly removing the test as well.

**Comment by taion** [comment]
fantastic, thanks!
